File: typescript/leetcodehard/wordLadder.ts
/*
https://leetcode.com/problems/word-ladder/

status - completed
*/

import { logp } from "../utils/logger";

function ladderLength(beginWord: string, endWord: string, wordList: string[]): number {
  const words = new Set(wordList);
  if (!words.has(endWord)) return 0;

  const queue: Array<[string, number]> = [[beginWord, 1]];
  const visited = new Set<string>([beginWord]);
  const alphabet = "abcdefghijklmnopqrstuvwxyz";

  while (queue.length > 0) {
    const [word, steps] = queue.shift()!;
    if (word === endWord) return steps;

    for (let i = 0; i < word.length; i++) {
      for (const ch of alphabet) {
        if (ch === word[i]) continue;
        const candidate = word.slice(0, i) + ch + word.slice(i + 1);
        if (words.has(candidate) && !visited.has(candidate)) {
          visited.add(candidate);
          queue.push([candidate, steps + 1]);
        }
      }
    }
  }

  return 0;
}

logp(
  `input hit->cog [hot,dot,dog,lot,log,cog] expected 5 actual ${ladderLength(
    "hit",
    "cog",
    ["hot", "dot", "dog", "lot", "log", "cog"],
  )}`,
);
logp(
  `input hit->cog [hot,dot,dog,lot,log] expected 0 actual ${ladderLength(
    "hit",
    "cog",
    ["hot", "dot", "dog", "lot", "log"],
  )}`,
);

File: com/dev/leetcodehard/WordLadder.java
/*
https://leetcode.com/problems/word-ladder/

status - completed
 */

package com.dev.leetcodehard;

import com.dev.logger.basePrinter;

import java.util.List;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;

class WordLadder extends basePrinter{

    static int ladderLength(String beginWord, String endWord, List<String> wordList){
        Set<String> dict = new HashSet<>(wordList);
        if (!dict.contains(endWord)){
            return 0;
        }

        Queue<String> queue = new LinkedList<>();
        queue.add(beginWord);
        Set<String> visited = new HashSet<>();
        visited.add(beginWord);

        int level = 1;

        while (!queue.isEmpty()){
            int size = queue.size();
            for (int i = 0; i < size; i++){
                String word = queue.poll();
                if (word.equals(endWord)){
                    return level;
                }

                char[] chars = word.toCharArray();
                for (int pos = 0; pos < chars.length; pos++){
                    char original = chars[pos];
                    for (char c = 'a'; c <= 'z'; c++){
                        if (c == original) continue;
                        chars[pos] = c;
                        String next = new String(chars);
                        if (dict.contains(next) && !visited.contains(next)){
                            visited.add(next);
                            queue.add(next);
                        }
                    }
                    chars[pos] = original;
                }
            }
            level++;
        }

        return 0;
    }

    public static void main(String[] args) {
        List<String> wordList1 = List.of("hot","dot","dog","lot","log","cog");
        logp("input hit->cog expected 5 actual " + ladderLength("hit", "cog", wordList1));

        List<String> wordList2 = List.of("hot","dot","dog","lot","log");
        logp("input hit->cog (no cog in dict) expected 0 actual " + ladderLength("hit", "cog", wordList2));

        List<String> wordList3 = List.of("a","b","c");
        logp("input a->c expected 2 actual " + ladderLength("a", "c", wordList3));
    }
}
